File: Java/IslandPerimeter.java
public class IslandPerimeter {
    public static int islandPerimeter(int[][] grid) {
        int pm = 0;
        int row = grid.length;
        int col = grid[0].length;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (grid[i][j] == 1) {
                    pm += 4;
                    if (j > 0 && grid[i][j - 1] == 1) pm -= 2;
                    if (i > 0 && grid[i - 1][j] == 1) pm -= 2;
                }
            }
        }
        return pm;
    }

    public static void main(String args[]) {
        System.out.println(islandPerimeter(new int[][]{{0, 1, 0, 0}, {1, 1, 1, 0}, {0, 1, 0, 0}, {1, 1, 0, 0}}));
    }
}

File: SwiftCode/IslandPerimeter.swift
//
//  IslandPerimeter.swift
//  SwiftCode
//
//  Created by Luminoid Li on 1/6/23.
//

import Foundation

//@main
private class Solution {
    // Time complexity: O(mn)
    // Space complexity: O(1)
    
    func islandPerimeter(_ grid: [[Int]]) -> Int {
        var perimeter = 0
        for (i, line) in grid.enumerated() {
            for (j, cell) in line.enumerated() {
                if cell == 1 {
                    perimeter += 4
                    if (i + 1 < grid.count) && grid[i + 1][j] == 1 {
                        perimeter -= 2
                    }
                    if (j + 1 < line.count) && grid[i][j + 1] == 1 {
                        perimeter -= 2
                    }
                }
            }
        }
        return perimeter
    }
    
    static func main() {
        let solution = Solution()
        print(solution.islandPerimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]))  // 16
        print(solution.islandPerimeter([[1]]))                                      // 4
        print(solution.islandPerimeter([[1,0]]))                                    // 4
    }
}
